docs/design/proposals/stateless-csi-plugin.md

```diff
@@ -0,0 +1,507 @@
+1. We should generate the same *ImageName*, when a Create request comes in with
+  the same *volume_name*
+  - CreateVolume may retry with the same *volume_name* on timeouts of the
+    request, and hence we need to generate the same *ImageName* for a given
+    *volume_name*
+1. We should be able to distinguish *ImageName*s that come from an instance of
+  the CSI plugin
+  - When dealing with a common Ceph cluster shared across CO systems, we need
+    the *ImageName* to be distinguishable, to avoid collisions and also to
+    be able support/aid operations like ListVolumes/Snapshots
+  - The same also applies when dealing with different CSI plugin instance in
+    the same CO system, using the same Ceph cluster
+
+#### Using Ceph OMaps for idempotent RPCs
+
+To preserve the idempotent CreateVolume RPC requirement, across *ImageName* for
+a given *volume_name*, it is proposed to use a per-CSI instance Ceph omap named
+`csi.volumes.[csi-id]`, to store keys named `csi.volume.[volume_name]` with the
+CO generated *volume_name*.
+
+The value stored in `csi.volume.[volume_name]` is the *ImageName* that is to be
+used for the requested *volume_name*.
+
+A create volume request would hence operate as follows,
+
+- Use CO passed *volume_name* to generate the required key name
+- Check if key exists and read the value of the same, if it does
+- If key exists and value points to an image that satisfies the create request
+  return with success
+- If not, generate an *ImageName* and store the key and value into the omap
+- Proceed with image creation operations as required
+
+Racy key creations/accesses are prevented by locking on the *volume_name* that
+the create request carries, thus ensuring that only one such transaction for
+the key named with the *volume_name* is in flight at any given time.
+
+One other advantage of using this omap is, list operations can be performed
+more easily using the omap, rather than listing all images in the pool or the FS.
+
+NOTE: For RBD plugin the RADOS objects and omaps would be stored in the same
+pool as the image
+
+NOTE: For CephFS plugin the RADOS objects would be stored in the metadata
+pool of the fsname passed to the create request. With further restrictions
+that the objects would be stored in a special `csicephfs` namespace on the
+metadata pool.
+
+#### Naming scheme
+
+ImageName format: `csi-vol-[uuid]`
+
+volume_id format: `[version]-[cluster-id]-[pool-id|fscid]-[uuid]`
+
+The spec states strings are a maximum of 128 bytes in length. Both
+*volume_name* and *volume_id* are strings, hence interpolating one into the
+other can break the size rules and rules out such encoding schemes.
+
+Due to the existence of the omap, we choose to create an ImageName that
+is independent of any implicit *volume_name* or *csi-id* encoding. The image
+once created would carry a `image-meta` key:value with, the key named
+`csi.volume.name` and the value containing the CO generated *volume_name*. This
```

Comment:
I'm confused. the value is the the ImageName, right? So the CO generated *volume_name* is same as the *ImageName*?

Reply:
This needs to be removed, it is a carry over from past iterations. There are no image meta keys stored on the image any further (and hence none stored on CephFS sub-directory as well)